discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py

```diff
@@ -0,0 +1,211 @@
+
+    def shutdown_all_nodes(self) -> None:
+        nodes = self.list_nodes()
+
+        for node in nodes:
+            self.shutdown_node(node.name)
+
+    def restart_node(self, node_name: str) -> None:
+        def reboot(vm) -> task:
+            return vm.ResetVM_Task()
+
+        self.__run_on_vm(node_name, self.cluster_name, reboot)
+
+    def format_node_disk(self, node_name: str, disk_index: int = 0) -> None:
+        raise NotImplementedError
+
+    def list_disks(self, node_name: str) -> List[Disk]:
+        raise NotImplementedError
+
+    def is_active(self, node_name) -> bool:
+        raise NotImplementedError
+
+    def format_all_node_disks(self) -> None:
+        raise NotImplementedError
+
+    def list_networks(self) -> List[Any]:
+        raise NotImplementedError
+
+    def list_leases(self, network_name: str) -> List[Any]:
+        raise NotImplementedError
+
+    def attach_test_disk(self, node_name: str, disk_size: int, bootable=False, persistent=False, with_wwn=False):
+        raise NotImplementedError
+
+    def detach_all_test_disks(self, node_name: str):
+        raise NotImplementedError
+
+    def get_ingress_and_api_vips(self) -> dict:
+        raise NotImplementedError
+
+    def get_cluster_network(self) -> str:
+        raise NotImplementedError
+
+    def setup_time(self) -> str:
+        raise NotImplementedError
+
+    def set_boot_order(self, node_name, cd_first=False) -> None:
+        raise NotImplementedError
+
+    def set_per_device_boot_order(self, node_name, key: Callable[[Disk], int]) -> None:
+        raise NotImplementedError
+
+    def get_host_id(self, node_name: str) -> str:
+        raise NotImplementedError
+
+    def set_cpu_cores(self, node_name: str, core_count: int) -> None:
+        raise NotImplementedError
+
+    def set_ram_kib(self, node_name: str, ram_kib: int) -> None:
+        raise NotImplementedError
+
+    def get_machine_cidr(self) -> str:
+        raise NotImplementedError
+
+    def attach_interface(self, node_name, network_xml: str) -> Tuple[libvirt.virNetwork, str]:
+        raise NotImplementedError
+
+    def add_interface(self, node_name, network_name, target_interface: str) -> str:
+        raise NotImplementedError
+
+    def undefine_interface(self, node_name: str, mac: str):
+        raise NotImplementedError
+
+    def create_network(self, network_xml: str) -> libvirt.virNetwork:
+        raise NotImplementedError
+
+    def get_network_by_name(self, network_name: str) -> libvirt.virNetwork:
+        raise NotImplementedError
+
+    def destroy_network(self, network: libvirt.virNetwork):
+        raise NotImplementedError
+
+    def set_single_node_ip(self, ip):
+        raise NotImplementedError
+
+    def __get_vm(self, node_name):
+        return next((vm for vm in self.__get_vms() if vm["attributes"]["name"] == node_name), None)
+
+    def __get_vms(self):
+        vms_object_type = self._tf.get_resources(resource_type="vsphere_virtual_machine")
```

Comment:
Change the annotations of `get_resources` to something more specific. 
At the moment it's just list but you addressing it as List[Dict[string, something]]